### ISSUE TITLE
New Rule Proposal: Warn against accessing a destructured property from object rest

### ISSUE THREAD

**Issue by HenriBeck** [eslint/eslint#10042]
**Please describe what the rule should do:**
Warn against trying to access a destructured property from an object rest.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Example 1
function someFunction({ disabled, ...rest }) {
  if (rest.disabled) { // Should warn here
    return true;
  }
}

// Example 2
const { a: { b }, ...rest } = someObject;

console.log(rest.a.b); // Should warn here

// Example 3
const { prop: someOtherName, ...rest } = someOtherObject;

console.log(rest.prop); // Should warn here, we could also provide the destructured variable name for the accessed property
```

**Why should this rule be included in ESLint (instead of a plugin)?**
This is a possible error which is framework / libary independent.
Of course, this rule would only be used when the user turned on the experimental object rest spread.

**Comment by ilyavolodin** [comment]
I think it's a reasonable rule to add, however, I think nested destructuring is going to be tough to track.

**Comment by HenriBeck** [comment]
Well if you destructure a nested object, it will remove the whole object from the rest object.
So it should be enough to check for the name of the nested destructured object.

```js
const obj = {
  a: { 
    b: true,
    d: false
  },
  c: false,
};

const { a: { b }, ...rest } = obj;

console.log(rest); // The rest object will only have the key 'c'
```

In this case, we would just need to check for `rest.a`.

**Comment by platinumazure** [comment]
This seems easy enough if the rest destructuring is at top level, but what about if the rest destructuring itself is nested?

```js
const obj = {
  a: { 
    b: true,
    d: false
  },
  c: false,
};

const { a: { b, ...rest } } = obj;

console.log(rest); // The rest object will only have the key 'd'?
```

**Comment by HenriBeck** [comment]
Well, we would need the keys where the object rest spread is being used but I think that should be possible.

**Comment by not-an-aardvark** [comment]
Is it necessary to know what keys the object has? It seems like we could just disallow accessing `b` from the object in the example from https://github.com/eslint/eslint/issues/10042#issuecomment-369916051, even if we don't know what the other keys are. That would work fine with nested destructuring (we would just disallow accessing all the keys that are neighbors of the rest element).

**Comment by HenriBeck** [comment]
@not-an-aardvark Yeah we only need the destructured keys/properties

**Comment by HenriBeck** [comment]
So this is something I could work on and create a PR?

**Comment by not-an-aardvark** [comment]
@HenriBeck In general, we accept rule proposals after there is a team member willing to champion the proposal, as well as three team members that leave a :+1: on the original issue. This issue hasn't been accepted yet, so we wouldn't end up merging the PR until that happens. You're certainly free to work on this and create a PR in the meantime, although there is a possibility that you would end up having wasted your time if the team decides not to accept this proposal.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is a strong consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
